app/server/api/token_api.py

```diff
@@ -46,11 +53,55 @@ def post(self):
 
             return make_response(jsonify(response_object)), 400
 
-        token = Token(address=address, name=name, symbol=symbol)
+        if deploy_smart_token_contract:
+
+            deploying_address = g.user.transfer_account.blockchain_address
+
+            if address:
+                response_object = {
+                    'message': "Must not supply address if deploying smart token"
+                }
+
+                return make_response(jsonify(response_object)), 400
+
+            if not exchange_contract_id:
+                response_object = {
+                    'message': 'Must supply exchange contract id if deploying smart token contract'
+                }
+
+                return make_response(jsonify(response_object)), 400
+
+            exchange_contract = ExchangeContract.query.get(exchange_contract_id)
+
+            if not exchange_contract:
+                response_object = {
+                    'message': 'Exchange contract not found for id {}'.format(exchange_contract_id)
+                }
+
+                return make_response(jsonify(response_object)), 400
+
+            smart_token_result = bt.deploy_smart_token(
+                deploying_address=deploying_address,
+                name=name, symbol=symbol, decimals=decimals,
+                issue_amount_wei=issue_amount_wei,
+                contract_registry_address=exchange_contract.contract_registry_blockchain_address,
+                reserve_token_address=exchange_contract.reserve_token.address,
+                reserve_ratio_ppm=reserve_ratio_ppm
+            )
+
+            address = smart_token_result['smart_token_address']
+            subexchange_address = smart_token_result['subexchange_address']
 
+        token = Token(address=address, name=name, symbol=symbol)
         db.session.add(token)
         db.session.commit()
```

Comment:
this commit appears to be necessary for some reason, otherwise  `exchange_contract.add_token(token, subexchange_address, reserve_ratio_ppm)` fails. not sure why

Reply:
You can normally make all kinds of queries without needing to persist to db though, as it just checks the orm session...? Like you can paste this pretty much anywhere an it works:

```
        u = TransferUsage(name='Foobar')
        db.session.add(u)
        found_u = TransferUsage.query.filter_by(name='Foobar').first()
```

ur right, should probably be a flush